netbox_onboarding/netbox_keeper.py

```diff
@@ -0,0 +1,373 @@
+
+            if not self.netdev_nb_platform_slug:
+                raise OnboardException(
+                    reason="fail-config",
+                    message=f"ERROR device platform not found: {self.netdev_hostname}"
+                )
+
+            self.nb_platform = Platform.objects.get(
+                slug=self.netdev_nb_platform_slug
+            )
+
+            logging.info(f"PLATFORM: found in NetBox {self.netdev_nb_platform_slug}")
+
+        except Platform.DoesNotExist:
+            if create_platform_if_missing:
+                self.nb_platform = Platform.objects.create(
+                    name=self.netdev_nb_platform_slug,
+                    slug=self.netdev_nb_platform_slug,
+                    napalm_driver=NETMIKO_TO_NAPALM[self.netdev_netmiko_device_type]
+                )
+            else:
+                raise OnboardException(
+                    reason="fail-general",
+                    message=f"ERROR platform not found in NetBox: {self.netdev_nb_platform_slug}"
+                )
+
+    def ensure_device_instance(self, default_status=PLUGIN_SETTINGS["default_device_status"]):
+        """Ensure that the device instance exists in NetBox and is assigned the provided device role or DEFAULT_ROLE.
+
+        Args:
+          default_status (str) : status assigned to a new device by default.
+        """
+
+        # Lookup if the device already exists in the NetBox
+        # First update and creation lookup is by checking the IP address
+        # of the onboarded device.
+        #
+        # If the device with a given IP is already in NetBox,
+        # any attributes including name could be updated
+        onboarded_device = None
+
+        try:
+            if self.netdev_mgmt_ip_address:
+                onboarded_device = Device.objects.get(
+                    primary_ip4__address__net_host=self.netdev_mgmt_ip_address
+                )
+        except Device.DoesNotExist:
+            logging.info(f"No devices found with primary-ip IP: "
+                         f"{self.netdev_mgmt_ip_address}")
+        except Device.MultipleObjectsReturned:
+            raise OnboardException(
+                reason='fail-general',
+                message=f"ERROR multiple devices using same IP in NetBox: "
+                        f"{self.netdev_mgmt_ip_address}"
+            )
+
+        if onboarded_device:
+            logging.info(f"Found device with primary-ip IP: {onboarded_device.name}")
+            lookup_args = {'pk': onboarded_device.pk,
+                           'defaults': dict(
+                               name=self.netdev_hostname,
+                               device_type=self.nb_device_type,
+                               device_role=self.nb_device_role,
+                               platform=self.nb_platform,
+                               site=self.nb_site,
+                               serial=self.netdev_serial_number
```

Comment:
Why no `status=default_status` in this case, unlike below where it's present?

Reply:
No need to update status for existing devices in case of re-onboarding process